Update Tier 3 HCR, and code for setting TACs w/ 1year projection

chengmatt opened the pull request:
1) Update how Tier 3 HCR was being computed,
2) Update how TACs should be set (1 year projection),
3) Add threshold cap control ruel

Note that I made a change in afscOM, so we need to commit those changes first, before this can be pulled through (just allowing for the output of an F vector). 

Also, lines 281 - 283 in sablefish_OM.R should be updated, where the selectivities should be weighted by the F in the terminal year and then summed and standardized - I didn't update since I didn't have enough time, but will make an issue as well.

Ovec8hkin committed:
Wrapper function for computing F35, F40, and B40 at once

Ovec8hkin committed:
Constant F HCR implementation

Ovec8hkin committed:
test: Extend sablefish OM to run multiple random simulations

Updated sablefish example to run multiple simulations and save resulting data to file

Ovec8hkin committed:
feat: Compute age structure summary statistics

Support for shannon diversity index, average age, proportion mature, proportion fully mature, and ABI

chengmatt committed:
Update Tier 3 HCR, and code for setting TACs w/ 1year projection

chengmatt committed:
Update Tier 3 HCR, and code for setting TACs w/ 1year projection

chengmatt committed:
add capped control rule

Ovec8hkin commented on R/sablefish_om.r:
```diff
@@ -295,26 +304,48 @@ for(y in 1:(nyears-1)){
+        proj_N <- array(NA, dim = c(dim(out_vars$naa_tmp)))
+        jointsel <- rbind(joint_self, joint_selm)
+    
+        for(s in 1:nsexes) {
+          for(a in 1:(nages - 1)) {
+            if(a == 1) proj_N[,1,s,] <- mean(recruitment) * 0.5 # using this for now, but assessment uses 1979 - terminal
+            else proj_N[,a,s,] <- out_vars$naa_tmp[,a,s,] * exp((out_f[y] * jointsel[s,a]) + M)
+          } # end a
+          proj_N[1,nages,s,1] <- out_vars$naa_tmp[1,nages-1,s,1]*exp(-(out_f[y] * jointsel[s,nages-1]) + M) +
+                                 out_vars$naa_tmp[1,nages,s,1]*exp(-(out_f[y] * jointsel[s,nages]) + M)
```
Let's think about how to refactor this to directly use `simulate_population`. It should be the same I think.

chengmatt replied:
Yeah, I did this pretty crudely, but i figured you could repackage it into these functions your wrote, which I'm not as familiar with.

Ovec8hkin commented on R/sablefish_om.r:
```diff
@@ -295,26 +304,48 @@ for(y in 1:(nyears-1)){
+          TAC_tmp <- 0
+          FAA_tmp <- (hcr_F[y] * jointsel[s,])
+          ZAA_tmp <- FAA_tmp + M
+          TAC_tmp <- sum(FAA_tmp / ZAA_tmp * proj_N[,,s,] * (1 - exp(-ZAA_tmp)) * dp.y$waa[,,s,])
```
This is the baranov equation right? If so, let's call `baranov()` directly here.

Ovec8hkin commented on R/sablefish_om.r:
```diff
@@ -3,9 +3,14 @@ rm(list=ls())
 library(devtools)
-devtools::load_all("~/Desktop/Projects/afscOM")
+library(patchwork)
+library(tidyverse)
+
+devtools::load_all("~/Desktop/Side projects/afscOM")
+source("~/Desktop/Side projects/SablefishMSE/R/reference_points.R")
+source("~/Desktop/Side projects/SablefishMSE/R/harvest_control_rules.R")
 
-assessment <- dget("~/Desktop/Projects/afscOM/data/test.rdat")
+assessment <- dget("~/Desktop/Side projects/afscOM/data/test.rdat")
```
I do this because I forget about it, but let's revert the changed pathing. The `library` calls and `source` calls can stay, but move to to using relative paths when sourcing stuff.

Ovec8hkin commented on R/sablefish_om.r:
```diff
@@ -295,26 +304,48 @@ for(y in 1:(nyears-1)){
+        proj_N <- array(NA, dim = c(dim(out_vars$naa_tmp)))
+        jointsel <- rbind(joint_self, joint_selm)
+    
+        for(s in 1:nsexes) {
```
Should do our best to figure out how to do this across both sexes simultaneously rather than looping.

Ovec8hkin commented on R/harvest_control_rules.R:
```diff
@@ -1,12 +1,8 @@
-    }else if(lrp < ssb && ssb < urp){
-        F <- fmax * (ssb-lrp)/(urp-lrp)
-    }
+npfmc_tier3_F <- function(ssb, b40, f40, alpha){
```
Let's add Roxygen docs to this at this point. I was initially passing `ssb/b40` to the function directly rather than as separate parameters, but it doesn't matter. But I'll get confused as to what the input is supposed to be soon...

Ovec8hkin reviewed (changes requested):
I'll work on some of the comments made here soon...

chengmatt committed:
update dimension for F

chengmatt committed:
update dimension for F

chengmatt committed:
roxygen for hcr

Ovec8hkin committed:
test: Analyses of age structure metrics

Analysis script for age structure metric, unit, and reference levels analyses. Plots used for UW Think Tank.

Ovec8hkin committed:
refactor: :art: Wrapped reference points calculation in function

Moved calculations of F35, F40, and B40 into a wrapper function (`calculate_ref_points`)

Ovec8hkin committed:
refactor: :art: Refactor TAC generation process to `simulate_TAC` function

Moved TAC generation/projection process to separate function and refectored to use `afscOM::simulate_population` and `afscOM::baranov` functions

Ovec8hkin committed:
Small updates

Ovec8hkin committed:
Merge branch 'main' into Tier3HCR+TAC

Ovec8hkin commented:
@chengmatt Check to make sure this runs correctly on your end

Ovec8hkin committed:
docs: Roxygen comments for HCR functions

Ovec8hkin committed:
docs: Roxygen for reference point calculations

chengmatt committed:
fix typos
